### PR TITLE
feat: return parameters only for DB with default driver

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This PR improves the `/api/v1/database/available/` endpoint:

1. For each DB engine spec, also show the list of installed drivers (eg, `mysqldb` and `mysqlconnector` for MySQL, in my system).
2. For each DB, show the default driver if available (eg, `mysqldb` for MySQL).
3. Only show the parameters if the default driver is installed (eg, if `mysqldb` is not installed but `mysqlconector` is, don't show the parameters for MySQL).

The reason for (3) is that if the user doesn't have the default driver installed we can't enable SSL for them. For example, for MySQL the engine spec can only enable it for the default `mysqldb` driver ([see](https://github.com/apache/superset/blob/master/superset/db_engine_specs/mysql.py#L66)).

If the user wants to use the non-default driver they need to use the old method via the SQLAlchemy URI. In that case we don't show the `parameters` for the engine spec, so that the modal should only allow the DB to be configured via the URI.

(1) and (2) will not be used by the modal, but are useful for debugging and troubleshooting.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

Updated unit tests.

Also tested manually. On my machine, with:

```python
PREFERRED_DATABASES = ["PostgreSQL", "Presto"]
```

The response in `/api/v1/database/available/` is:

```json
{
  "databases": [
    {
      "available_drivers": [
        "mysqlconnector", 
        "mysqldb"
      ], 
      "default_driver": "mysqldb", 
      "engine": "mysql", 
      "name": "MySQL", 
      "parameters": {
        "properties": {
          "database": {
            "description": "Database name", 
            "type": "string"
          }, 
          "encryption": {
            "description": "Use an encrypted connection to the database", 
            "type": "boolean"
          }, 
          "host": {
            "description": "Hostname or IP address", 
            "type": "string"
          }, 
          "password": {
            "description": "Password", 
            "nullable": true, 
            "type": "string"
          }, 
          "port": {
            "description": "Database port", 
            "format": "int32", 
            "type": "integer"
          }, 
          "query": {
            "additionalProperties": {}, 
            "description": "Additional parameters", 
            "type": "object"
          }, 
          "username": {
            "description": "Username", 
            "nullable": true, 
            "type": "string"
          }
        }, 
        "required": [
          "database", 
          "host", 
          "port", 
          "username"
        ], 
        "type": "object"
      }, 
      "preferred": false, 
      "sqlalchemy_uri_placeholder": "mysql://user:password@host:port/dbname[?key=value&key=value...]"
    }, 
    {
      "available_drivers": [
        "psycopg2"
      ], 
      "default_driver": "psycopg2", 
      "engine": "postgresql", 
      "name": "PostgreSQL", 
      "parameters": {
        "properties": {
          "database": {
            "description": "Database name", 
            "type": "string"
          }, 
          "encryption": {
            "description": "Use an encrypted connection to the database", 
            "type": "boolean"
          }, 
          "host": {
            "description": "Hostname or IP address", 
            "type": "string"
          }, 
          "password": {
            "description": "Password", 
            "nullable": true, 
            "type": "string"
          }, 
          "port": {
            "description": "Database port", 
            "format": "int32", 
            "type": "integer"
          }, 
          "query": {
            "additionalProperties": {}, 
            "description": "Additional parameters", 
            "type": "object"
          }, 
          "username": {
            "description": "Username", 
            "nullable": true, 
            "type": "string"
          }
        }, 
        "required": [
          "database", 
          "host", 
          "port", 
          "username"
        ], 
        "type": "object"
      }, 
      "preferred": true, 
      "sqlalchemy_uri_placeholder": "postgresql://user:password@host:port/dbname[?key=value&key=value...]"
    }, 
    {
      "available_drivers": [
        "pysqlite"
      ], 
      "engine": "sqlite", 
      "name": "SQLite", 
      "preferred": false
    }, 
    {
      "available_drivers": [
        "vertica_python"
      ], 
      "engine": "vertica", 
      "name": "Vertica", 
      "preferred": false
    }, 
    {
      "available_drivers": [
        "rest"
      ], 
      "engine": "presto", 
      "name": "Presto", 
      "preferred": true
    }, 
    {
      "available_drivers": [
        "rest"
      ], 
      "engine": "druid", 
      "name": "Apache Druid", 
      "preferred": false
    }, 
    {
      "available_drivers": [
        "bigquery"
      ], 
      "default_driver": "bigquery", 
      "engine": "bigquery", 
      "name": "Google BigQuery", 
      "parameters": {
        "properties": {
          "credentials_info": {
            "description": "Contents of BigQuery JSON credentials.", 
            "type": "string", 
            "x-encrypted-extra": true
          }
        }, 
        "type": "object"
      }, 
      "preferred": false, 
      "sqlalchemy_uri_placeholder": "bigquery://{project_id}"
    }, 
    {
      "available_drivers": [
        "turbodbc"
      ], 
      "engine": "mssql", 
      "name": "Microsoft SQL", 
      "preferred": false
    }, 
    {
      "available_drivers": [
        "apsw"
      ], 
      "engine": "gsheets", 
      "name": "Google Sheets", 
      "preferred": false
    }
  ]
}
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/databases/api.py [hunk 1]
```diff
@@ -920,13 +939,25 @@ def available(self) -> Response:
 
             available_databases.append(payload)
 
-        available_databases.sort(
-            key=lambda payload: preferred_databases.index(payload["engine"])
-            if payload["engine"] in preferred_databases
-            else len(preferred_databases)
```
Comment (betodealmeida) [inline superset/databases/api.py:926]: This was sorting the preferred DBs correctly, but doing a non-deterministic sort on the non-preferred. 🤦

**Comment by eschutho** [comment]
> The reason for (3) is that if the user doesn't have the default driver installed we can't enable SSL for them. For example, for MySQL the engine spec can only enable it for the default mysqldb driver (see).

Is SSL the only difference between the default and non default driver? Could we show them the form but not give them the option to turn on ssl (like we do for admins who enforce ssl but the opposite)?

**Comment by betodealmeida** [comment]
> > The reason for (3) is that if the user doesn't have the default driver installed we can't enable SSL for them. For example, for MySQL the engine spec can only enable it for the default mysqldb driver (see).
> 
> Is SSL the only difference between the default and non default driver? Could we show them the form but not give them the option to turn on ssl (like we do for admins who enforce ssl but the opposite)?

Right, but there could potentially be other parameters that are driver-dependent — maybe a MySQL driver uses client certificates instead of username/password — so I think we should only show the form if the default driver is installed.

Also, if we show the form when the default driver is not installed, and the user has more than one non-default drivers installed, this raises the problem of which one we should use. Eg, the default driver for MySQL is `mysqldb`, but the user could have `oursql` and `mysqlconnector` installed instead. If we show them the form for MySQL, we would have to choose one of the drivers for them (normally we would use the default driver).

I think users in cases like this probably know better, and so they can just the SQLAlchemy URI to configure the DB with a non-default driver.

**Review by hughhhh — APPROVED** [review]
LGTM

**Pull request merged**